rdma_link_linux.go

```diff
@@ -143,3 +143,128 @@ func (h *Handle) RdmaLinkSetName(link *RdmaLink, name string) error {
 
 	return execRdmaSetLink(req)
 }
+
+func netnsModeToString(mode uint8) string {
+	netnsModes := map[uint8]string{0: "exclusive", 1: "shared"}
+
+	if mode > (uint8)(len(netnsModes)-1) {
+		return "unknown"
+	}
+	return netnsModes[mode]
+}
+
+func executeOneGetRdmaNetnsMode(data []byte) (string, error) {
+	var modeResult string
+
+	reader := bytes.NewReader(data)
+	for reader.Len() >= 4 {
+		_, attrType, len, value := parseNfAttrTLV(reader)
+
+		switch attrType {
+		case nl.RDMA_NLDEV_SYS_ATTR_NETNS_MODE:
+			var mode uint8
+			r := bytes.NewReader(value)
+			binary.Read(r, nl.NativeEndian(), &mode)
+			modeResult = netnsModeToString(mode)
+			return modeResult, nil
+		}
+		if (len % 4) != 0 {
+			// Skip pad bytes
+			reader.Seek(int64(4-(len%4)), seekCurrent)
+		}
+	}
+	return modeResult, fmt.Errorf("Invalid netns mode")
+}
+
+// RdmaSystemGetNetnsMode gets the net namespace mode for RDMA subsystem
+// Returns mode string and error status as nil on success or returns error
+// otherwise.
+// Equivalent to: `rdma system show netns'
+func RdmaSystemGetNetnsMode() (string, error) {
+	return pkgHandle.RdmaSystemGetNetnsMode()
+}
+
+// RdmaSystemGetNetnsMode gets the net namespace mode for RDMA subsystem
+// Returns mode string and error status as nil on success or returns error
+// otherwise.
+// Equivalent to: `rdma system show netns'
+func (h *Handle) RdmaSystemGetNetnsMode() (string, error) {
+	var modeResult string
+
+	proto := getProtoField(nl.RDMA_NL_NLDEV, nl.RDMA_NLDEV_CMD_SYS_GET)
+	req := h.newNetlinkRequest(proto, unix.NLM_F_ACK)
+
+	msgs, err := req.Execute(unix.NETLINK_RDMA, 0)
+	if err != nil {
+		return modeResult, err
+	}
+	if len(msgs) == 0 {
+		return modeResult, fmt.Errorf("No valid response from kernel")
+	}
+	modeResult, err = executeOneGetRdmaNetnsMode(msgs[0])
+	if err != nil {
+		return modeResult, err
+	}
+	return modeResult, nil
+}
+
+func netnsModeStringToUint8(mode string) (uint8, error) {
+	netnsModes := map[string]uint8{"exclusive": 0, "shared": 1}
+
+	val, found := netnsModes[mode]
+	if !found {
+		return 0, fmt.Errorf("Invalid mode")
```

Comment:
please print the mode # for debuggability `[...]mode: %q", mode)`

Reply:
@aboch done